### PR TITLE
User defined network segmentation l3/l2 foundation: pod2pod

### PR DESCRIPTION
**Pull request by tssurya**
#### Which issue(s) this PR fixes

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: dist/images/ovnkube.sh [hunk 1]
```diff
@@ -1203,6 +1205,12 @@ ovn-master() {
   fi
   echo "multi_network_enabled_flag=${multi_network_enabled_flag}"
 
+  network_segmentation_enabled_flag=
+  if [[ ${ovn_network_segmentation_enable} == "true" ]]; then
+	  network_segmentation_enabled_flag="--enable-multi-network --enable-network-segmentation"
+  fi
+  echo "network_segmentation_enabled_flag=${network_segmentation_enabled_flag}"
```
Comment (qinqon) [inline dist/images/ovnkube.sh:1212]: We have to also add ${network_segmentation_enabled_flag}  to `  /usr/bin/ovnkube --init-ovnkube-controller`  and friends.

Comment (tssurya) [reply dist/images/ovnkube.sh:1212]: good catch quique!

File: go-controller/pkg/config/config.go [hunk 2]
```diff
@@ -1020,6 +1021,12 @@ var OVNK8sFeatureFlags = []cli.Flag{
 		Destination: &cliConfig.OVNKubernetesFeature.EnableMultiNetworkPolicy,
 		Value:       OVNKubernetesFeature.EnableMultiNetworkPolicy,
 	},
+	&cli.BoolFlag{
```
Comment (tssurya) [inline go-controller/pkg/config/config.go:1024]: TODO: Add some unit tests for this flag

File: go-controller/pkg/util/multi_network.go [hunk 3]
```diff
@@ -530,6 +530,10 @@ func IsMultiNetworkPoliciesSupportEnabled() bool {
 	return config.OVNKubernetesFeature.EnableMultiNetwork && config.OVNKubernetesFeature.EnableMultiNetworkPolicy
 }
 
+func IsNetworkSegmentationSupportEnabled() bool {
+	return config.OVNKubernetesFeature.EnableMultiNetwork && config.OVNKubernetesFeature.EnableNetworkSegmentation
```
Comment (qinqon) [inline go-controller/pkg/util/multi_network.go:534]: We should return false if we are not at IC, since this feature is only support for it.

**Review by qinqon — CHANGES_REQUESTED** [review]

**Comment by qinqon** [comment]
```
I0613 05:33:50.857824      16 kube.go:154] Setting annotations map[k8s.ovn.org/active-network:hllj2_tenant-blue] on namespace primary-network-5868
E0613 05:33:50.859357      16 kube.go:163] Error in setting annotation on namespace primary-network-5868: namespaces "primary-network-5868" is forbidden: User "system:serviceaccount:ovn-kubernetes:ovnkube-cluster-manager" cannot patch resource "namespaces/status" in API group "" in the namespace "primary-network-5868"
I0613 05:33:50.859414      16 network_attach_def_controller.go:239] cluster-manager: Finished syncing net-attach-def primary-network-5868/tenant-blue: 1.870971ms
I0613 05:33:50.859462      16 network_attach_def_controller.go:272] "Error syncing net-attach-def, retrying" net-attach-def="primary-network-5868/tenant-blue" err="failed ensuring namespace active network: failed annotating namespace 'primary-network-5868' with the active network 'hllj2_tenant-blue'"
```

File: go-controller/pkg/util/multi_network.go (hunk 3 above)
Comment (tssurya) [reply go-controller/pkg/util/multi_network.go:534]: @trozet wdyt? shall we support this feature for only nonIC?

File: go-controller/pkg/allocator/pod/pod_annotation.go [hunk 4]
```diff
@@ -265,6 +265,10 @@ func allocatePodAnnotationWithRollback(
 		IPs:      podAnnotation.IPs,
 		MAC:      podAnnotation.MAC,
 		TunnelID: podAnnotation.TunnelID,
+		Primary:  false,
```
Comment (tssurya) [inline go-controller/pkg/allocator/pod/pod_annotation.go:268]: is this path expected to be hit during default net info path?

File: go-controller/pkg/config/config.go (hunk 2 above)
Comment (tssurya) [reply go-controller/pkg/config/config.go:1024]: done!

File: go-controller/pkg/allocator/pod/pod_annotation.go (hunk 4 above)
Comment (tssurya) [reply go-controller/pkg/allocator/pod/pod_annotation.go:268]: spoke to Jaime (tl;dr currently we don't call this path from default netinfo but in future we could; so better to have that contingency in place)

File: go-controller/pkg/allocator/pod/pod_annotation.go [hunk 5]
```diff
@@ -265,6 +275,25 @@ func allocatePodAnnotationWithRollback(
...
+			}
+			activeNetworkForPod := util.GetNamespaceActiveNetwork(podNamespace)
+			if activeNetworkForPod == nadName {
+				tentative.Primary = true
+			}
+		}
+	} else if !netInfo.IsSecondary() {
+		// if user defined network segmentation is not enabled
+		// then we know pod's primary network is "default"
+		tentative.Primary = true
```
Comment (tssurya) [inline go-controller/pkg/allocator/pod/pod_annotation.go:296]: check if pods on upgrade should have this new field reflected in them? generally speaking any pod annotation with no changes will not have this reflected.

Comment (tssurya) [reply go-controller/pkg/allocator/pod/pod_annotation.go:296]: upgrades should be tested for this. perhaps even pod annotations needs to be behind feature gate if we were to start using this field in 4.16 etc? UGH.

Comment (tssurya) [reply go-controller/pkg/allocator/pod/pod_annotation.go:296]: hmm maybe its just better to put the pod annotation also behind a feature gate...

**Comment by tssurya** [comment]
```
2024-06-17T19:22:08.7857533Z [90m------------------------------[0m
2024-06-17T19:22:08.7858093Z [0mGateway unit tests[0m [90mgetGatewayNextHops[0m [0mIn Local GW mode[0m 
2024-06-17T19:22:08.7858917Z   [1mFinds correct gateway interface and nexthops when dummy gateway bridge is created[0m
2024-06-17T19:22:08.7860335Z   [37m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/testing/testing.go:18[0m
2024-06-17T19:22:08.7862131Z W0617 19:22:08.785907   28538 exec.go:265] Unexpected command: /fake-bin/ovs-vsctl --timeout=15 port-to-br br-ex
2024-06-17T19:22:08.7863394Z 
2024-06-17T19:22:08.7863682Z Executed commands do not match expected commands!
2024-06-17T19:22:08.7864563Z Executed unexpected command /fake-bin/ovs-vsctl --timeout=15 port-to-br br-ex
2024-06-17T19:22:08.7864963Z 
2024-06-17T19:22:08.7865200Z [00] ! /fake-bin/ovs-vsctl --timeout=15 port-to-br enf1f0
2024-06-17T19:22:08.7865607Z goroutine 53 [running]:
2024-06-17T19:22:08.7865878Z runtime/debug.Stack()
2024-06-17T19:22:08.7866382Z 	/opt/hostedtoolcache/go/1.21.11/x64/src/runtime/debug/stack.go:24 +0x67
2024-06-17T19:22:08.7867143Z runtime/debug.PrintStack()
2024-06-17T19:22:08.7867632Z 	/opt/hostedtoolcache/go/1.21.11/x64/src/runtime/debug/stack.go:16 +0x1d
2024-06-17T19:22:08.7868701Z github.com/ovn-org/ovn-kubernetes/go-controller/pkg/testing.(*FakeExec).Command(0xc0003cb6e0, {0xc000058960, 0x13}, {0xc0004d8b70, 0x3, 0x3})
2024-06-17T19:22:08.7869827Z 	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/testing/exec.go:266 +0xdac
2024-06-17T19:22:08.7870807Z github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util.run({0xc000058960, 0x13}, {0xc0004d8b70, 0x3, 0x3})
2024-06-17T19:22:08.7871789Z 	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/util/ovs.go:299 +0xb1
2024-06-17T19:22:08.7872723Z github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util.RunOVSVsctl({0xc000652da8, 0x2, 0xc000652cd8?})
2024-06-17T19:22:08.7873842Z 	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/util/ovs.go:318 +0x1bf
2024-06-17T19:22:08.7875092Z github.com/ovn-org/ovn-kubernetes/go-controller/pkg/node.getGatewayNextHops()
2024-06-17T19:22:08.7876684Z 	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/gateway_init.go:168 +0xb7b
2024-06-17T19:22:08.7877743Z github.com/ovn-org/ovn-kubernetes/go-controller/pkg/node.glob..func6.6.5.1()
2024-06-17T19:22:08.7878683Z 	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/gateway_init_linux_test.go:1828 +0xdd7
2024-06-17T19:22:08.7879523Z github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc000177800)
2024-06-17T19:22:08.7880685Z 	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113 +0x12b
2024-06-17T19:22:08.7882318Z github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0xc000177800)
2024-06-17T19:22:08.7884182Z 	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:64 +0x138
2024-06-17T19:22:08.7885437Z github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run(0xc000120180)
2024-06-17T19:22:08.7886518Z 	/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/it_node.go:26 +0x95
2024-06-17T19:22:08.7887909Z github.com/onsi/ginkgo/internal/spec.(*Spec).runSample(0xc000448ff0, 0x54?, {0x38c25c0?, 0xc000352440})
```
ERR ...

File: go-controller/pkg/allocator/pod/pod_annotation.go [hunk 6]
```diff
@@ -265,6 +275,25 @@ func allocatePodAnnotationWithRollback(
...
+	if util.IsNetworkSegmentationSupportEnabled() {
+		tentative.Primary = netInfo.IsSecondary() && netInfo.IsPrimaryNetwork()
+		if !netInfo.IsSecondary() { // only do this extra bit if needed
+			var podNamespace *v1.Namespace
+			podNamespace, err = namespaceLister.Get(pod.Namespace)
+			if err != nil {
+				return
+			}
+			activeNetworkForPod := util.GetNamespaceActiveNetwork(podNamespace)
+			if activeNetworkForPod == nadName {
```
Comment (qinqon) [inline go-controller/pkg/allocator/pod/pod_annotation.go:289]: This should be compared with network not nadName

**Review by qinqon — CHANGES_REQUESTED** [review]

**Comment by tssurya** [comment]
https://github.com/ovn-org/ovn-kubernetes/pull/4462 will superceed this, keeping this PR in draft mode. Will close eventually.

**Pull request closed without merging**